src/main/java/hu/oe/nik/szfmv/automatedcar/systemcomponents/PowertrainSystem.java

```diff
@@ -15,17 +35,229 @@
+
+                if (brakePedalPosition == 0) {
+                    LOGGER.debug(":: doPowertrain() method called: Slowing down to minimum speed");
+                    // Acceleration
+                    if (speedDelta < 0 && (speed > CarSpecifications.MAX_REVERSE_SPEED)) {
+                        adjustSpeed(speedDelta);
+                    }
+                    // Enginebreak
+                    if (speedDelta > 0 && speed < CarSpecifications.MIN_REVERSE_SPEED) {
+                        adjustSpeed(speedDelta);
+                    }
+
+
+                } else {
+                    LOGGER.debug(":: doPowertrain() method called: Braking, allow to stop to zero");
+                    if (speed < 0) {
+                        adjustSpeed(speedDelta);
+                    }
+                    if (speed > 0) {
+                        speed = 0;
+                        powertrainPacket.setSpeed(speed);
+                    }
+                }
+
+                break;
+
+            case D:
+                orientationVector = 1;
+                shiftLevel = 1;
+                gearShiftWatcher(speedDelta);
+
+                if (brakePedalPosition == 0) {
+                    LOGGER.debug(":: doPowertrain() method called: Accelerate to maximum speed");
+                    // Acceleration
+                    if (speedDelta > 0 && speed < CarSpecifications.MAX_FORWARD_SPEED) {
+                        adjustSpeed(speedDelta);
+                    }
+                    // Enginebrake
+                    if (speedDelta < 0 && speed > CarSpecifications.MIN_FORWARD_SPEED) {
+                        adjustSpeed(speedDelta);
+                    }
+                } else {
+                    LOGGER.debug(":: doPowertrain() method called: Braking, allow to stop to zero");
+                    if (speed > 0) {
+                        adjustSpeed(speedDelta);
+                    }
+                    if (speed < 0) {
+                        speed = 0;
+                        powertrainPacket.setSpeed(speed);
+                    }
+                }
+                break;
+
+            default:
+                break;
+        }
     }
 
-    public double getSpeed() {
-        return this.speed;
+    /**
+     * Adjust new speed
+     *
+     * @param speedDelta speed delta
+     */
+    private void adjustSpeed(double speedDelta) {
+        speed += speedDelta;
+        powertrainPacket.setSpeed(speed);
+    }
+
+    @Override
+    public void getVirtualFunctionBusSignals() {
+        try {
+            gasPedalPosition = virtualFunctionBus.inputPacket.getGasPedalPosition();
+            brakePedalPosition = virtualFunctionBus.inputPacket.getBreakPedalPosition();
+            gearState = virtualFunctionBus.inputPacket.getGearState();
+        } catch (NullPointerException ex) {
+            LOGGER.error(ex);
+            throw new NullPointerException("Inputpacket was not initiated when reading signals:\n" + ex.toString());
```

Comment:
Azzal kezeled le a kivételt, hogy újra kiváltod??? Ezt így nem tudom értelmezni.
Főleg, hogy emiatt kb. leáll a program.
Ezt úgy tudnám elképzelni, hogy definiáljátok a pl. NotInitializedPacketException-t, amit egy null check (if inputPacket is null...) után kiváltotok.

Reply:
A @SzFMV2018-Tavasz/team3 -tól kódformázási javításokat kértem, amint ezeket sikerül leküzdeni elvileg bekerül a masterbe a várt packet.

Reply:
@pintergreg 
1. Sajnos az a béna kivétel kezelés úgy maradt, nem akartam így hagyni. Javítom a javaslat szerint, köszönöm. Ez a customexception hova mehet? Milyen packet lenne a célszerű?
2. Igen, még nincs megoldása az inputpacket-nek, kiderült hogy az a @SzFMV2018-Tavasz/team3 hatásköre, egyik kommentben megjelöltem Őket, de még nem jött infó.

Tanácsot szeretnék kérni, ilyenkor mi az eljárás? A használt featurebranch-ben javítom a fenti dolgokat, majd PR team5-be és akkor látszik a master PR-ben is a változás?

Reply:
Még egy kérdésem volna, ha dobok a null check után egy NotInitializedPacketException, akkor az hol lesz kezelve?
Az AutomatedCar osztályig el kell dobni és ott kezeljem? Elég ha csak logolja, de a játék ne álljon le?

Reply:
Mivel a packetek buszhoz kötődnek szerintem a hu/oe/nik/szfmv/automatedcar/bus-ba simán jó helyen van egy ilyen.

Akár általánosra is csinálhatod: Simán MissingPacketException-re vagy valami hasonló, csomagtípus függetlenül és akkor akár más is közvetlenül használhatja.

Az AutomatedCarban elvileg jónak tűnik. Ezen még kéne kicsit gondolkodni, szóval nem kőbevésett igazság.

Reply:
Köszönöm, elkészült benne is van a team5 branchben. Már látszik itt is. Így rendben lesz?